File: Backend/EventCompetitionService/src/main/java/tn/matchmakers/eventcompetitionservice/services/serviceInterfaces/EventService.java
package tn.matchmakers.eventcompetitionservice.services.serviceInterfaces;

import tn.matchmakers.eventcompetitionservice.dto.CreateEventRequest;
import tn.matchmakers.eventcompetitionservice.dto.EventResponseDto;
import tn.matchmakers.eventcompetitionservice.dto.UpdateEventRequest;
import tn.matchmakers.eventcompetitionservice.dto.external.MatchDto;
import tn.matchmakers.eventcompetitionservice.dto.external.SportDto;
import tn.matchmakers.eventcompetitionservice.entities.Event;
import tn.matchmakers.eventcompetitionservice.entities.enums.StatutEvent;

import java.time.LocalDate;
import java.util.List;

public interface EventService {
    // ─── CRUD
    EventResponseDto createEvent(CreateEventRequest dto, String token);
    List<Event> getAll();
    Event getById(String id);
    EventResponseDto update(String id, UpdateEventRequest dto, String token);
    void delete(String id);
    // ─── Métier
    Event changeStatut(String id, StatutEvent statut);
    Event assignCompetition(String eventId, String competitionId);
    List<Event> getByStatut(StatutEvent statut);
    List<Event> getByOrganizer(String userId);
    List<Event> getByDateRange(LocalDate from, LocalDate to);
    // ─── Appels externes
    SportDto getSportDetails(String sportId);
    List<MatchDto> getMatchesForCompetition(String competitionId);
    EventResponseDto joinTeam(String eventId, String teamId, String token);
    EventResponseDto leaveTeam(String eventId, String teamId, String token);
    EventResponseDto joinEvent(String eventId, String token);
    EventResponseDto leaveEvent(String eventId, String token);

    // ─── Localisation
    List<Event> getByLocation(String city);
    List<Event> getByLocationAndStatut(String city, StatutEvent statut);

}

File: Frontend/src/app/features/events/service/event.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Event, StatutEvent, CreateEventRequest, UpdateEventRequest, EventType } from '../../../features/events/event.model';

@Injectable({
  providedIn: 'root'
})
export class EventService {

  private readonly API = 'http://localhost:8083/eventsCompetitions/api/events';
  private readonly TYPES_API = 'http://localhost:8083/eventsCompetitions/api/event-types';

  constructor(private http: HttpClient) {}

  //Crud
  getAll(): Observable<Event[]> {
    return this.http.get<Event[]>(this.API);
  }
  getById(id: string): Observable<Event> {
    return this.http.get<Event>(`${this.API}/${id}`);
  }
  create(dto: CreateEventRequest): Observable<Event> {
    return this.http.post<Event>(this.API, dto);
  }
  update(id: string, dto: UpdateEventRequest): Observable<Event> {
    return this.http.put<Event>(`${this.API}/${id}`, dto);
  }
  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API}/${id}`);
  }

  //Filtres
  getByStatut(statut: StatutEvent): Observable<Event[]> {
    return this.http.get<Event[]>(`${this.API}/statut/${statut}`);
  }
  getByLocation(city: string): Observable<Event[]> {
    return this.http.get<Event[]>(`${this.API}/location`, {
      params: new HttpParams().set('city', city)
    });
  }
  getByLocationPlanned(city: string): Observable<Event[]> {
    return this.http.get<Event[]>(`${this.API}/location/planned`, {
      params: new HttpParams().set('city', city)
    });
  }
  getByOrganizer(userId: string): Observable<Event[]> {
    return this.http.get<Event[]>(`${this.API}/organizer/${userId}`);
  }
  // Statut
  changeStatut(id: string, statut: StatutEvent): Observable<Event> {
    return this.http.patch<Event>(`${this.API}/${id}/statut`, null, {
      params: new HttpParams().set('statut', statut)
    });
  }
  // Équipes 
  joinTeam(eventId: string, teamId: string): Observable<Event> {
    return this.http.post<Event>(`${this.API}/${eventId}/teams/${teamId}`, {});
  }
  leaveTeam(eventId: string, teamId: string): Observable<Event> {
    return this.http.delete<Event>(`${this.API}/${eventId}/teams/${teamId}`);
  }
  // Participants individuels
  joinEvent(id: string): Observable<Event> {
    return this.http.post<Event>(`${this.API}/${id}/join`, {});
  }
  leaveEvent(id: string): Observable<Event> {
    return this.http.delete<Event>(`${this.API}/${id}/leave`);
  }
  // Event Types
  getEventTypes(): Observable<EventType[]> {
    return this.http.get<EventType[]>(this.TYPES_API);
  }
  createEventType(type: Partial<EventType>): Observable<EventType> {
    return this.http.post<EventType>(this.TYPES_API, type);
  }

  deleteEventType(id: string): Observable<void> {
    return this.http.delete<void>(`${this.TYPES_API}/${id}`);
  }
}